File: unity_code/Gameplay/RocketStation.cs
using Assets.Scripts.Exceptions;
using UnityEngine;
using Assets.Scripts.Actions;
using System;
using Random = UnityEngine.Random;
using System.Linq;

namespace Assets.Scripts.Gameplay
{
    [RequireComponent(typeof(Transform))]
    public class RocketStation : MonoBehaviour
    {
        private const int MaxDelayMilliseconds = 300;

        public GameObject PipePrototype;
        public GameObject RocketPrototype;
        public GameActionExecutor ActionExecutor;
        public Game GameStats;

        public int RocketStationHeight = 10;
        public int RocketStationWidth = 10;
        public int PipesSpawnOffset = 15;
        public int PipesFallingTime = 10;

        private Pipe[,] rocketStationField;
        private Rocket[] rockets;

        private void Start ()
        {
            rocketStationField = new Pipe[RocketStationHeight, RocketStationWidth];
            rockets = new Rocket[RocketStationHeight];
            CheckParameters();
            GenerateRockets ();
            GameStats.GameEnd += Stop;
            ActionExecutor.AllActionsEnd += TryExplodePipes;
            GeneratePipes();
        }

        private void CheckParameters()
        {
            if (PipePrototype == null)
                throw new ParameterRequiredException("PipePrototype");

            if (ActionExecutor == null)
                throw new ParameterRequiredException("ActionDirector");

            if (RocketPrototype == null)
                throw new ParameterRequiredException("RocketPrototype");

            if (GameStats == null)
                throw new ParameterRequiredException("GameStats");
        }

        private void GenerateRockets()
        {
            var column = RocketStationWidth;
            for (var i = 0; i < RocketStationHeight; i++)
            {
                if(rockets[i] != null)
                    continue;

                rockets[i] = Instantiate(RocketPrototype, GetTilePosition(i, column), Quaternion.identity).GetComponent<Rocket>();
            }
        }

        private void Stop()
        {
            ActionExecutor.AllActionsEnd -= TryExplodePipes;
            ActionExecutor.Stop();

            for (var i = 0; i < RocketStationHeight; i++)
                for (var j = 0; j < RocketStationWidth; j++)
                {
                    if(rocketStationField[i, j] == null)
                        continue;

                    rocketStationField[i, j].Pressed -= OnPipeTilePressed;
                    rocketStationField[i, j].Kill();
                }

            foreach(var rocket in rockets)
                rocket.Kill();
        }

        private void TryExplodePipes()
        {
            if(ActionExecutor.IsRunning)
                return;

            LaunchRockets();
            ShufflePipes();
            GeneratePipes();
        }

        private void LaunchRockets()
        {
            GameStats.AddScores(rockets.Where(r => r.ReadyToLaunch).ToArray());

            for (var i = 0; i < RocketStationHeight; i++)
                rockets[i].Launch();
        }

        private void GeneratePipes()
        {
            FillRocketStationWithPipes();
            CalculatePipeConnections();
        }

        private void CalculatePipeConnections()
        {
            ConnectPipes();
            StartSteam();
            StartRockets();
        }

        private void FillRocketStationWithPipes()
        {
            for (var i = 0; i < RocketStationHeight; i++)
                for (var j = 0; j < RocketStationWidth; j++)
                {
                    if(rocketStationField[i, j] != null)
                        continue;

                    rocketStationField[i, j] = Instantiate(PipePrototype, GetInitialTilePosition(i, j), Quaternion.identity).GetComponent<Pipe>();
                    rocketStationField[i, j].Pressed += OnPipeTilePressed;

                    MovePipe(i, j);
                }
        }

        private void OnPipeTilePressed()
        {
            CalculatePipeConnections();
            TryExplodePipes();
        }

        private void MovePipe(int row, int column, int? destination = null)
        {
            var destinationRow = destination ?? row;
            ActionExecutor.Run(new SequenceAction(
                new DelayAction(TimeSpan.FromMilliseconds(Random.Range(0, MaxDelayMilliseconds))),
                new MoveAction(rocketStationField[row, column], GetTilePosition(destinationRow, column), TimeSpan.FromSeconds(PipesFallingTime))));
        }

        private Vector3 GetInitialTilePosition(int row, int column)
        {
            var position = GetTilePosition(row, column);
            position.y += PipesSpawnOffset;
            return position;
        }

        private Vector3 GetTilePosition(int row, int column)
        {
            var tileSize = PipePrototype.GetComponent<BoxCollider2D>().size;
            return new Vector3(column * tileSize.x + transform.position.x + tileSize.x / 2, - row * tileSize.y + transform.position.y - tileSize.y / 2);
        }

        private void StartSteam()
        {
            for (var i = 0; i < RocketStationHeight; i++)
            {
                var pipe = rocketStationField[i, 0];
                if (pipe.IsSideConnected(PipeSide.Left))
                    pipe.ConnectSteam();
            }
        }

        private void StartRockets()
        {
            for (var i = 0; i < RocketStationHeight; i++)
            {
                var pipe = rocketStationField[i, RocketStationWidth - 1];
                if(pipe.IsSideConnected(PipeSide.Right))
                {
                    pipe.ConnectRocket();

                    if(pipe.IsReadyToLaunch)
                        rockets[i].PrepareForLaunch();
                }
            }
        }

        private void ShufflePipes()
        {
            for (var j = 0; j < RocketStationWidth; j++)
                ShuffleColumn(j);
        }

        private void ShuffleColumn(int column)
        {
            var emptySlot = -1;
            for (var i = RocketStationHeight - 1; i >= 0; i--)
            {
                if (rocketStationField[i, column].IsReadyToLaunch)
                {
                    DestroyPipe(i, column);
                    emptySlot = FindEmptySlot(i, column, emptySlot);
                }
                else if(emptySlot != -1)
                {
                    MovePipe(i, column, emptySlot);
                    AdjustField(i, column, emptySlot);
                    emptySlot = FindEmptySlot(i, column, emptySlot);
                }
            }
        }

        private void DestroyPipe(int row, int column)
        {
            rocketStationField[row, column].Pressed -= TryExplodePipes;
            var pipe = rocketStationField[row, column];
            rocketStationField[row, column] = null;
            pipe.Kill();
        }

        private int FindEmptySlot(int row, int column, int emptySlot)
        {
            if(emptySlot == -1)
                return row;

            for (var i = emptySlot; i >= 0; i--)
                if(rocketStationField[i, column] == null)
                    return i;

            throw new SlotNotFoundException(row, column);
        }

        private void AdjustField(int row, int column, int emptySlot)
        {
            rocketStationField[emptySlot, column] = rocketStationField[row, column];
            rocketStationField[row, column] = null;
        }

        private void ConnectPipes()
        {
            for (var j = 0; j < RocketStationWidth; j++)
                for (int i = 0; i < RocketStationHeight; i++)
                    ConnectPipe(i, j);
        }

        private void ConnectPipe(int row, int column)
        {
            var pipe = rocketStationField[row, column];
            pipe.Reset();
            foreach (var side in pipe.PipeConnectionSides)
                HandleSide(side, pipe, row, column);
        }

        private void HandleSide(PipeSide side, Pipe pipe, int row, int column)
        {
            switch (side)
            {
                case PipeSide.Up:
                    pipe.UpPipe = GetUpPipe(row, column);
                    break;
                case PipeSide.Right:
                    pipe.RightPipe = GetRightPipe(row, column);
                    break;
                case PipeSide.Down:
                    pipe.DownPipe = GetDownPipe(row, column);
                    break;
                case PipeSide.Left:
                    pipe.LeftPipe = GetLeftPipe(row, column);
                    break;
            }
        }

        private Pipe GetUpPipe(int row, int column)
        {
            return row == 0 || !rocketStationField[row - 1, column].IsSideConnected(PipeSide.Down) ? null : rocketStationField[row - 1, column];
        }

        private Pipe GetRightPipe(int row, int column)
        {
            return column == RocketStationWidth - 1 || !rocketStationField[row, column + 1].IsSideConnected(PipeSide.Left) ? null : rocketStationField[row, column + 1];
        }

        private Pipe GetDownPipe(int row, int column)
        {
            return row == RocketStationHeight - 1 || !rocketStationField[row + 1, column].IsSideConnected(PipeSide.Up) ? null : rocketStationField[row + 1, column];
        }

        private Pipe GetLeftPipe(int row, int column)
        {
            return column == 0 || !rocketStationField[row, column - 1].IsSideConnected(PipeSide.Right) ? null : rocketStationField[row, column - 1];
        }
    }
}

File: src/main/ts/RocketStation.ts
/// <reference path="../../../lib/phaser/phaser.d.ts"/>

import { Pipe } from './Pipe';
import { Rocket } from './Rocket';
import { Config } from './Config';
import { PipeSide } from './PipeSide';
import {Rockets} from "./Rockets";
import { MoveAction } from './actions/MoveAction';
import { DelayAction } from './actions/DelayAction';
import { SequenceAction } from './actions/SequenceAction';
import { GameActionExecutor } from './actions/GameActionExecutor';

import TileDimensions = Config.TileDimensions;

export class RocketStation {
    private rocketStationField:Pipe[][];
    private rockets:Rocket[];

    private game:Phaser.Game;
    private position:Phaser.Point;
    private rocketGame:Rockets;
    private actionExecutor:GameActionExecutor;

    constructor(game:Phaser.Game, position:Phaser.Point, rocketGame:Rockets) {
        this.rockets = [];
        this.rocketStationField = [];
        this.position = position;
        this.game = game;
        this.rocketGame = rocketGame;

        this.actionExecutor = new GameActionExecutor();
        Rockets.addAnimatable(this.actionExecutor);

        this.generateRockets();

        this.actionExecutor.allActionsEnd.add(this.tryExplodePipes, this);

        this.generatePipes();
        this.tryExplodePipes();
    }

    private generateRockets():void {
        let column = Config.RocketStationParameters.ROCKET_STATION_WIDTH;
        for(let i = 0; i < Config.RocketStationParameters.ROCKET_STATION_HEIGHT; i++) {
            if(this.rockets[i] != null) {
                continue;
            }

            let rocketPosition = this.getTilePosition(i, column);
            rocketPosition.x += Config.RocketStationParameters.ROCKET_OFFSET;
            
            this.rockets[i] = new Rocket(this.game, rocketPosition);
        }
    }

    private getTilePosition(row:number, column:number):Phaser.Point {
        return new Phaser.Point(
            column * Config.TileDimensions.WIDTH + this.position.x + Config.TileDimensions.WIDTH / 2,
            row * Config.TileDimensions.HEIGHT - this.position.y + Config.TileDimensions.HEIGHT / 2)
    }

    private generatePipes():void {
        this.fillRocketStationWithPipes();
        this.calculatePipeConnections();
    }

    private fillRocketStationWithPipes():void {
        for(let i = 0; i < Config.RocketStationParameters.ROCKET_STATION_HEIGHT; i++) {
            for(let j = 0; j < Config.RocketStationParameters.ROCKET_STATION_WIDTH; j++) {
                if(this.rocketStationField[i] == null) {
                    this.rocketStationField[i] = [];
                }

                if(this.rocketStationField[i][j] != null) {
                    continue;
                }

                this.rocketStationField[i][j] = new Pipe(this.game, this.getInitialTilePosition(i, j));
                this.rocketStationField[i][j].pressed.add(this.onPipeTilePressed, this);

                this.movePipe(i, j, i);
            }
        }
    }

    private onPipeTilePressed():void {
        this.calculatePipeConnections();
        this.tryExplodePipes();
    }

    private getInitialTilePosition(row:number, column:number):Phaser.Point {
        let point = this.getTilePosition(row, column);
        point.y = -2 * TileDimensions.HEIGHT;
        return point;
    }

    private movePipe(row:number, column:number, destinationRow:number) {
        let destination = this.getTilePosition(destinationRow, column);
        this.actionExecutor.run(new SequenceAction([
            new DelayAction(200),
            new MoveAction(this.rocketStationField[row][column], destination, 300)]));
    }

    private calculatePipeConnections():void {
        this.connectPipes();
        this.startSteam();
        this.startRockets();
    }

    private connectPipes():void {
        for(let j = 0; j < Config.RocketStationParameters.ROCKET_STATION_WIDTH; j++) {
            for(let i = 0; i < Config.RocketStationParameters.ROCKET_STATION_HEIGHT; i ++) {
                this.connectPipe(i, j);
            }
        }
    }

    private connectPipe(row:number, column:number):void {
        let pipe = this.rocketStationField[row][column];
        pipe.reset();
        for(let side of pipe.getPipeConnectionSides()) {
            this.handleSide(side, pipe, row, column);
        }
    }

    private handleSide(side:PipeSide, pipe:Pipe, row:number, column:number):void {
        switch(side) {
            case PipeSide.Up: {
                pipe.upPipe = this.getUpPipe(row, column);
                return;
            }
            case PipeSide.Right: {
                pipe.rightPipe = this.getRightPipe(row, column);
                return;
            }
            case PipeSide.Down: {
                pipe.downPipe = this.getDownPipe(row, column);
                return;
            }
            case PipeSide.Left: {
                pipe.leftPipe = this.getLeftPipe(row, column);
                return;
            }
        }
    }

    private getUpPipe(row:number, column:number):Pipe {
        return row == 0 || !this.rocketStationField[row - 1][column].isSideConnected(PipeSide.Down) ? null : this.rocketStationField[row - 1][column];
    }

    private getRightPipe(row:number, column:number):Pipe {
        return column == Config.RocketStationParameters.ROCKET_STATION_WIDTH - 1 || !this.rocketStationField[row][column + 1].isSideConnected(PipeSide.Left) ? null : this.rocketStationField[row][column + 1];
    }

    private getDownPipe(row:number, column:number):Pipe {
        return row == Config.RocketStationParameters.ROCKET_STATION_HEIGHT - 1 || !this.rocketStationField[row + 1][column].isSideConnected(PipeSide.Up) ? null : this.rocketStationField[row + 1][column];
    }

    private getLeftPipe(row:number, column:number):Pipe {
        return column == 0 || !this.rocketStationField[row][column - 1].isSideConnected(PipeSide.Right) ? null : this.rocketStationField[row][column - 1];
    }

    private startSteam():void {
        for(let i = 0; i < Config.RocketStationParameters.ROCKET_STATION_HEIGHT; i++) {
            let pipe = this.rocketStationField[i][0];
            if(pipe.isSideConnected(PipeSide.Left)) {
                pipe.connectSteam();
            }
        }
    }

    private startRockets():void {
        for(let i = 0; i < Config.RocketStationParameters.ROCKET_STATION_HEIGHT; i++) {
            let pipe = this.rocketStationField[i][Config.RocketStationParameters.ROCKET_STATION_WIDTH - 1];
            if(pipe.isSideConnected(PipeSide.Right)) {
                pipe.connectRocket();

                if(pipe.isReadyToLaunch()) {
                    this.rockets[i].prepareForLaunch();
                }
            }
        }
    }

    private tryExplodePipes():void {
        if(this.actionExecutor.isRunning)
            return;

        this.launchRockets();
        this.shufflePipes();
        this.generatePipes();
    }

    private launchRockets():void {
        const launchedRockers:Array<Rocket> = [];
        for (var i = 0; i < Config.RocketStationParameters.ROCKET_STATION_HEIGHT; i++) {
            let rocket = this.rockets[i];
            if(rocket.isReadyToLaunch()) {
                launchedRockers.push(rocket);
                this.rockets[i].launch();
            }
        }
        this.rocketGame.addScores(launchedRockers);
    }

    private shufflePipes():void {
        for (var j = 0; j < Config.RocketStationParameters.ROCKET_STATION_WIDTH; j++) {
            this.shuffleColumn(j);
        }
    }

    private shuffleColumn(column:number):void {
        var emptySlot = -1;
        for (var i = Config.RocketStationParameters.ROCKET_STATION_HEIGHT - 1; i >= 0; i--) {
            if (this.rocketStationField[i][column].isReadyToLaunch()) {
                this.destroyPipe(i, column);
                emptySlot = this.findEmptySlot(i, column, emptySlot);
            }
            else if(emptySlot != -1) {
                this.movePipe(i, column, emptySlot);
                this.adjustField(i, column, emptySlot);
                emptySlot = this.findEmptySlot(i, column, emptySlot);
            }
        }
    }

    private destroyPipe(row:number, column:number):void {
        this.rocketStationField[row][column].pressed.add(this.tryExplodePipes, this);
        var pipe = this.rocketStationField[row][column];
        this.rocketStationField[row][column] = null;
        pipe.kill();
    }

    private findEmptySlot(row:number, column:number, emptySlot:number):number {
        if(emptySlot == -1) {
            return row;
        }

        for (var i = emptySlot; i >= 0; i--) {
            if(this.rocketStationField[i][column] == null) {
                return i;
            }
        }

        throw "Empty slot not found.";
    }

    private adjustField(row:number, column:number, emptySlot:number):void {
        this.rocketStationField[emptySlot][column] = this.rocketStationField[row][column];
        this.rocketStationField[row][column] = null;
    }
}